### ISSUE TITLE
Live Migration

### ISSUE THREAD

**Issue by yisun-git** [cloud-hypervisor/cloud-hypervisor#75]
Crate Name
==========================================================================
vm-live-migration

Short Description
==========================================================================
vm-live-migration crate abstracts the components used for live migration,
provides the interfaces to control the live migration and defines the flows
of live migration.

The upper layer which can respond to control requests calls vm-live-migration
interface to start the live migration. The user input configuration is passed
as parameters.

Live migration is a complex process. Many methods/components have been invented
to help to improve the migration. At the beginning phase, only are the methods/
components required to make basic live migration work implemented.

Why is this crate relevant to the rust-vmm project?
==========================================================================
Live migration is one the major benefits of virtualization to provide
flexible management to virtual machine (VM). For energy management, load
balancing, server maintenance and fault tolerance reasons in cloud computing,
there is strong demand to migrate VM from one host machine to another with
minimal downtime. Live migration is the technology to fulfill the requirement.
It can move the running VM from one physical machine to another without
disconnecting the applications. It already is a must have feature in VMM.

Design
==========================================================================
### Overview
![image](https://user-images.githubusercontent.com/22880141/60107726-f653fa80-9799-11e9-9f4d-270a122c4dae.png)

There are source VM and target VM to do live migration. The flows of them are
different. So we need define a MigrationSender struct and a MigrationReceiver
struct to handle different works. But some common works can be abstracted to
MigrationBase struct which is included by both MigrationSender and
MigrationReceiver.

Ram is the most data to be transferred during live migration. Its flow is
somewhat complex compared with others. So MigrationRam struct is defined to
handle ram data migration.

Besides ram data, the VM states (including cpu, devices, migration configuration
and so on) should be migrated too. So MigrationState struct is defined to
provide interfaces to outside devices/components to register states need to do
live migration.

Both source side and target side should follow a fixed data format to recognize
the transferring data. So MigrationDataFile struct is defined to provide
interfaces to handle the data as the fixed format. When data is ready or
achieves threshold on source side, MigrationDataFile calls MigrationTransport
interface to send data to target. On target, when data is received, it is parsed
as the format defined by MigrationDataFile.

Live migration may base on different protocols, e.g. TCP, RDMA and so on. So we
need abstract a MigrationTransport trait to cover all common requirements for
data transportation. MigrationTCP struct is implemented as the concrete
implementation for MigrationTransport trait. Only is TCP planned to be supported
at this beginning phase.

Conclude above components relationships as below figures.

When source and target VMs start, the MigrationState instances are registered
into MigrationStateArray.
![image](https://user-images.githubusercontent.com/22880141/60107774-0d92e800-979a-11e9-8b65-cf751a22e55d.png)

When migration starts, the work flow of all components.
![image](https://user-images.githubusercontent.com/22880141/60107917-4b900c00-979a-11e9-9fab-9f51414636e7.png)

## MigrationTransport

MigrationTransport trait defines the transportation protocols and provides
related interfaces to establish connection, handling packages and so on. At
current phase, only is TCP supported as a concrete implementation.

```
pub trait MigrationTransport {
    pub fn bind(...) -> bool;
    pub fn connect(...) -> bool;
    pub fn shutdown(...) -> bool;
    pub fn write(...) -> i32;
    pub fn read(...) -> i32;
    ...
}
```

## MigrationTCP

MigrationTCP is the concrete implementation for MigrationTransport trait. It
manages TCP connection, sends/receives migration data. It interacts with mio
crate to call basic functions.

When migration starts, MigrationTCP instance is created if the configured
protocol is TCP. It is saved as a MigrationTransport instance in upper layer.

```
pub struct MigrationTCP {
    addr: String,
    sock: TcpStream,
    ...
}

impl MigrationTCP {
    pub fn new(addr: String, ...) -> MigrationTCP;
    ...
}

impl MigrationTransport for MigrationTCP {
    ...
}
```

## MigrationDataFile

MigrationDataFile defines the mechanism to handle the data. It provides a series
of interfaces to put/get the data from buffer.

```
pub struct MigrationDataFile {
    trans: Box<MigrationTransport>,
    buf_pos: u32,
    buf_size: u32,
    buf: [u8; 32768],
    iovcnt: u32,
    /* Part of buf[] is saved into one iov[] member */
    iov: [IoVec; 64],
}

impl MigrationDataFile {
    pub fn new(trans: Box<MigrationTransport>, ...) -> MigrationDataFile;
    pub fn put_byte(&self, v: u32);
    pub fn put_be16(&self, v: u32);
    pub fn put_be32(&self, v: u32);
    pub fn put_be64(&self, v: u64);
    pub fn put_buffer(&self, buf: &mut [u8], size: usize);
    pub fn get_byte(&self) -> u32;
    pub fn get_be16(&self) -> u32;
    pub fn get_be32(&self) -> u32;
    pub fn get_be64(&self) -> u64;
    pub fn get_buffer(&self, buf: &mut [u8], size: usize);
    /* Call MigrationTransport interface to send data to target */
    pub fn flush_data(...);
    ...
}
```

## MigrationDataFileFlag

To make source VM and target VM identify the content of transferring data, we
should define some data flags to mark what the following data is. Considering
the possibility to make live migration work between rust-vmm and Qemu, we
follow Qemu transferring data format.

So below flags are defined. They will be inserted into MigrationDataFile as data
header/footer or flag. The migration flow implementation will do this work.

```
pub const MIGRATION_VM_FILE_MAGIC: u32          = 0x5145564d;
pub const MIGRATION_VM_FILE_VERSION_COMPAT: u32 = 0x00000002;
pub const MIGRATION_VM_FILE_VERSION: u32        = 0x00000003;

pub const MIGRATION_VM_EOF: u8                  = 0x00;
pub const MIGRATION_VM_SECTION_START: u8        = 0x01;
pub const MIGRATION_VM_SECTION_PART: u8         = 0x02;
pub const MIGRATION_VM_SECTION_END: u8          = 0x03;
pub const MIGRATION_VM_SECTION_FULL: u8         = 0x04;
pub const MIGRATION_VM_SUBSECTION: u8           = 0x05;
pub const MIGRATION_VM_VMDESCRIPTION: u8        = 0x06;
pub const MIGRATION_VM_CONFIGURATION: u8        = 0x07;
pub const MIGRATION_VM_COMMAND: u8              = 0x08;
pub const MIGRATION_VM_SECTION_FOOTER: u8       = 0x7e;

pub const MIGRATION_RAM_SAVE_FLAG_ZERO: u8      = 0x02;
pub const MIGRATION_RAM_SAVE_FLAG_MEM_SIZE: u8  = 0x04;
pub const MIGRATION_RAM_SAVE_FLAG_PAGE: u8      = 0x08;
pub const MIGRATION_RAM_SAVE_FLAG_EOS: u8       = 0x10;
pub const MIGRATION_RAM_SAVE_FLAG_CONTINUE: u8  = 0x20;
```

## MigrationRam

Ram is the most data to be transferred during live migration. Its flow is
somewhat complex compared with others. It needs to interact with memory
management component to get dirty pages, send and restore them. To reduce the
time and size to transfer the ram data, there are many methods invented. So we
make ram migration handling a standalone component. If time allowed, the
compression method may be adopted as one of the features of the first
development phase.

Here is a prerequisite that memory management component should provide
interfaces to get dirty memory information.

```
pub struct DirtyRegion {
    /* The dirty memory region */
    region: MmapRegion,
    /* Dirty bitmap to record which page is dirty in this memory block */
    bit_map: Vec<u64>,
    num_dirty_pages: u64,
}

pub struct MigrationRam {
    mig_data: MigrationDataFile,
    /* Is Arc necessary? */
    dirty_regions: Arc<Vec<DirtyRegion>>,
    last_sent_region: DirtyRegion,
    ...
}

impl MigrationRam {
    pub fn new(...) -> MigrationRam;
    pub fn migrate_ram_send_prepare(...);
    pub fn migrate_ram_receive_prepare(...);
    /* Call MigrationDataFile interface to insert dirty ram */
    pub fn migrate_ram_send(...);
    /* Call MigrationDataFile interface to get dirty ram */
    pub fn migrate_ram_receive(...);
    pub fn migrate_ram_iterate(...);
    pub fn migrate_dirty_bitmap_sync(...);
    pub fn migrate_find_dirty_region(...);
    ...
}
```

## MigrationState

MigrationState defines the data structure used by any component which needs to
migrate its states, e.g. cpu/virtio-device. The MigrationStateField defines
the format of attribute to be migrated. Caller should set the attribute type
so that we can know how to parse it. The value itself is converted to u8 vector
to avoid complex implementation because the unknow type of value. Furthermore,
caller should register save_fn/load_fn for its specific work.

When vmm starts on both source side and target side, its components that need
to migrate state should call migration_state_push() to register its own
MigrationState instance into global vector MigrationStateArray.

When live migration starts, source side migration process iterates
MigrationStateArray to save states into MigrationDataFile buffer. When target side
receives the MigrationDataFile, it parses out the state section and iterates
MigrationStateArray to find the matched MigrationState instance according to
idstr and instance_id. Then, it set the content of MigrationDataFile into
MigrationState/MigrationStateField as sequence. At last, call load_fn callback
function to ask the component load the migrated state value.

```
/* Field types */
const FIELD_TYPE_BYTE   = 1 << 1;
const FIELD_TYPE_BOOL   = 1 << 2;
const FIELD_TYPE_U16    = 1 << 3;
const FIELD_TYPE_U32    = 1 << 4;
const FIELD_TYPE_U64    = 1 << 5;
/* E.g. U16 array can be "FIELD_TYPE_U16 | FIELD_TYPE_ARRAY" */
const FIELD_TYPE_ARRAY  = 1 << 6;
...

pub struct MigrationStateField {
    value: Vec<u8>,
    version: i32,
    /* According to type, call corresponding MigrationDataFile interface to handle data */
    type: u32,
    size: u32,
    /* Number of array members */
    num_array: u32,
}

/* Corresponding to Qemu "SaveStateEntry + VMStateDescription" */
pub struct MigrationState {
    /* The identifier to match MigrationDataFile with the correct handler */
    idstr: String,
    instance_id: i32,
    version: i32,
    fields: Vec<MigrationStateField>,
    /*
     * Devices should register their save/load callback functions when create
     * MigrationState instance. They are used for device specific purpose.
     */
    save_fn: Box<FnMut()>,
    load_fn: Box<FnMut()>,
}

impl MigrationState {
    pub fn new(...) -> MigrationState {...};
    pub fn send_state(...) {...};
    pub fn receive_state(...) {...};
    ...
}

lazy_static! {
    static ref MigrationStateArray: Mutex<Vec<MigrationState>> = Mutex::new(vec![]);
}

pub fn migration_state_push(state: MigrationState, ...) {
    MigrationStateArray.lock().unwrap().push(state);
    ...
}

pub fn migration_state_pop(...) -> MigrationState {
    ...
    MigrationStateArray.lock().unwrap().pop()
}
```

## MigrationBase

MigrationBase struct is the base class to manage migration status and provides
the interfaces to control migration process.

MigrationStatus is used to manage live migration status.

```
pub enum MigrationStatus {
    Migration_Status_Launch,
    Migration_Status_Init,
    Migration_Status_Active,
    Migration_Status_Cancel,
    Migration_Status_Complete,
    Migration_Status_Fail,
}

pub enum MigrationProtocol {
    Migration_Protocol_TCP,
    ...
}

pub struct MigrationBase {
    status: MigrationStatus,
    protocol: MigrationProtocol,
    data: MigrationDataFile,
    ram: MigrationRam,
    trans_thread: thread::JoinHandle<()>,
    ...
}

impl MigrationBase {
    pub fn new() -> Result<Self> {...};
    pub fn set_status() -> Result<()> {...};
    pub fn start_transport_thread() -> thread::JoinHandle<()> {...};
    ...
}
```

## MigrationSender

MigrationSender struct includes Migration struct object and implements the
sender specific data and operations.

```
pub struct MigrationSender {
    mig: MigrationBase,
    ...
}

impl MigrationSender {
    pub fn new() -> Result<Self> {...};
    pub fn start() -> Result<()> {...};
    pub fn cancel() -> Result<()> {...};
    pub fn send_configure() -> Result<()> {...};
    pub fn migrate_iterate() -> Result<()> {...};
    ...
}
```

Above functions are used to manage live migration on source side.
- start(): When user input command on source side to start live migration
  process, this function is called. The main flow of live migration is below.
    Stage 1: Mark all VM RAM as dirty to transfer all memory to destination to
             initialize the target VM.
    Stage 2: Keep sending dirty RAM pages since last iteration until a
             convergence criteria is fulfilled.
    Stage 3: Suspend VM and transfer remaining dirty RAM, CPU states and device
             states to destination. Furthermore, redirect source VM network
             traffic to target VM.
- cancel(): When user input command to cancel live migration process, this
  function is called.

## MigrationReceiver

MigrationReceiver struct includes MigrationBase struct object and implements the
receiver specific data and operations.

```
pub struct MigrationReceiver {
    mig: MigrationBase,
    ...
}

impl MigrationReceiver {
    pub fn new() -> Result<Self> {...};
    pub fn apply_config() -> Result<()> {...};
    pub fn receive_loop() -> Result<()> {...};
    ...
}
```

Above functions are used to manage live migration on target side.
- receive_loop(): When user input command on target side to start live migration
  process, MigrationReceiver instance is created to do some preparations. Then,
  this function is called to wait for source side message. When message comes,
  it is parsed and dispatched to different flows as the section header.
- apply_config(): It is used to apply configuration data received from source
  side.

**Comment by sameo** [comment]
I think this is ready to be sent to rust-vmm for a wider review there.
@rbradford @sboeuf @chao-p WDYT?

**Comment by chao-p** [comment]
Agree, we should post it to rust-vmm to get more reviews.

**Comment by sboeuf** [comment]
To be honest, I didn't get the time to look through it yet, but I agree it'd be nice to have feedback from the community rather sooner than later.

**Comment by yisun-git** [comment]
Ok, I will post it to rust-vmm. Thanks!

**Comment by yisun-git** [comment]
A crate addition request has been submitted on rust-vmm:
https://github.com/rust-vmm/community/issues/67

**Comment by sboeuf** [comment]
@yisun-git 
Following our conversation from this morning, the question was, why do we need `MigrationStateArray` to be static as defined below:
```rust
lazy_static! {
    static ref MigrationStateArray: Mutex<Vec<MigrationState>> = Mutex::new(vec![]);
}
```

My understanding so far is that we are trying to send a list of `MigrationState` structures (each specific to a component) to the target VMM so that it can recreate those states for the target VM.
Why can't we create the `MigrationStateArray` structure on the fly and provision it when needed?

Regarding the problem about not being able to use some references from the `Vm` structure, can you point out a concrete example?

My understanding is that it shouldn't matter how the `Vm` structure is defined, as long as we provide the correct implementation to copy/clone the relevant data held by this structure into the `MigrationStateArray` structure.

**Comment by yisun-git** [comment]
Thanks for comments!

> @yisun-git
> Following our conversation from this morning, the question was, why do we need `MigrationStateArray` to be static as defined below:
> 
> ```rust
> lazy_static! {
>     static ref MigrationStateArray: Mutex<Vec<MigrationState>> = Mutex::new(vec![]);
> }
> ```
> 
> My understanding so far is that we are trying to send a list of `MigrationState` structures (each specific to a component) to the target VMM so that it can recreate those states for the target VM.
> Why can't we create the `MigrationStateArray` structure on the fly and provision it when needed?
> 
Before migration thread starts, some struct instances are created, e.g. Vmm/Vm/Vcpu/Devices (migration thread is started by external command during source VM running). Migration thread cannot directly get these instances to get/load values for some fields. So we need a global static array to make all these instances can register itself into it. Then, migration thread can iterate the array to get/load values of fields.

How do we create MigrationStateArray on the fly? Per my thought, it seems only input the top level instance, i.e. Vmm instance, and make sure the Vmm instance can be accessed by multi-threads. This looks more difficult.

> Regarding the problem about not being able to use some references from the `Vm` structure, can you point out a concrete example?
> 
Vm is just for test. At the end, we may not need to migration its states. But there are some structs with lifetime annotation in cloud-hypervisor. So I think it is meaningful to investigate that. For test, I added one field, version: Vec<u8>, in Vm struct.
```
 918 pub struct Vm<'a> {
 919     fd: Arc<VmFd>,
 920     kernel: File,
 921     memory: GuestMemoryMmap,
 922     vcpus: Vec<thread::JoinHandle<()>>,
 923     devices: DeviceManager,
 924     cpuid: CpuId,
 925     config: VmConfig<'a>,
 926     epoll: EpollContext,
 927     on_tty: bool,
 928     version: Vec<u8>,
 929 }
```

> My understanding is that it shouldn't matter how the `Vm` structure is defined, as long as we provide the correct implementation to copy/clone the relevant data held by this structure into the `MigrationStateArray` structure.

Yes, your opinion is right. The tricky point is how to get the instance in migration thread to get/load fields values. My current design is to define a trait as part of MigrationState and make the migrated instance implement the trait, then register the instance into MigrationState. Per test, this mechanism works well for structs without lifetime annotation but fails for other case. I have found two possible workarounds to have a try.
```
  5 pub trait MigrationStateFn {
  6     fn save_fn(&self, data: Arc<MigrationDataFile>);
  7     fn load_fn(&mut self, data: Arc<MigrationDataFile>);
  8 }
  9
 10 pub struct MigrationState {
 11     pub idstr: &'static str,
 12     pub state_fn: Box<MigrationStateFn + Send>,
 13 }
```

**Comment by sboeuf** [comment]
@yisun-git 
Ok so there are two aspects in this discussion:

First one is about being able to share the same instance between multiple threads, since you expect the migration code to run in a dedicated thread. I think more generally, we're gonna need the VMM to run a dedicated thread for a server taking care of things like migration, hotplug, etc...
Now, based on the fact that we might need to pass those information to the new thread, we will need instances that implement `Send` and `Sync`. And this translates directly into using `Arc<>` (if an immutable reference is needed) or `Arc<Mutex<>>` (if a mutable reference is needed).
But one alternative to send all instances to the new thread would be to copy anything that needs to be copied before to send it to the new thread. That's where the migration specific code could invoke `Clone/Copy` implementation of the instance we need to copy. And this would be done before the new thread is created.

The second aspect is about lifetimes for traits. As long as the structure implementing a trait defines the lifetimes appropriately, Rust will allow you to pass a reference which needs a specific lifetime. And of course, as long as the lifetime of the instance you're trying to pass is longer than the lifetime of the structure implementing the trait, you should be fine.

This is a lot of theory, and having a full example to debug might be helpful.

**Comment by yisun-git** [comment]
Hi, Sebastien,

You analysis is correct. I upload my draft codes to [private repo](https://github.com/yisun-git/cloud-hypervisor_migration). Please check it.

Sorry for late because I was busy working on SIOV project.

**Comment by fazlamehrab** [comment]
I understand the presented structures which are necessary to represent VM state and for migration. However, I was wondering about a few things for which you already have thoughts of, I guess.

1. Described three different stages indicate how the data are going to be transferred. However, it is not clear what algorithm we are going to use for live-migration. Is it going to be pre-copy or post-copy live-migration?

2. At the destination machine, how the VM in the target machine is going to restore the states from the migration structures?

3. Can we leverage existing checkpoint/restart technique like CRIU for VM live-migration?

4. Do we have any plan to migrate between heterogeneous-ISAs like x86 and ARM?

**Comment by yisun-git** [comment]
> I understand the presented structures are necessary to represent VM state and for migration. However, I was wondering about a few other things which you might already have thought of, I guess.
> 
> 1. Described three different stages indicate how the data are going to be transferred. However, it is not clear what algorithm we are going to use for live-migration. Is it going to be pre-copy or post-copy live-migration?

Per plan, pre-copy first. Post-copy will be implemented after it.

> 2. At the destination machine, how the VM in the target machine is going to restore the states from the migration structures?

The states values are stored in MigrationState::fields. Through the callback function, load_fn, defined in 'trait MigrationStateFn', the values can be restored on target machine.

> 3. Can we leverage existing checkpoint/restart technique like CRIU for VM live-migration?

This is an interesting idea. In fact, pre-copy/post-copy defines some checkpoints and restore them. We may do some investigations on CRIU to see if anything can be leveraged.

> 4. Do we have any plan to migrate between heterogeneous-ISAs like x86 and ARM?

So far, we do not have such plan.

**Comment by yisun-git** [comment]
Hi,

I missed one source file to upload to the repo. Just fixed it. Please pull again if you have cloned it yesterday. Sorry for that.

**Comment by sboeuf** [comment]
@yisun-git 
>I missed one source file to upload to the repo. Just fixed it. Please pull again if you have cloned it yesterday. Sorry for that.

No worries :)

**Comment by fazlamehrab** [comment]
Thanks @yisun-git 

> Through the callback function, load_fn, defined in 'trait MigrationStateFn', the values can be restored on target machine.

Could you please share more detail on when the load_fn is going to be called and how this function is going to work? Is it like we are going to launch a new VM on the target machine with the similar configuration from the source machine, and stop the execution at some point to restore the VM state from  MigrationState::fields? Or, is it like the application memory will be unmapped first and load_fn will be called when there is a corresponding page fault?  Or, you have something else in your mind?

> This is an interesting idea. In fact, pre-copy/post-copy defines some checkpoints and restore them. We may do some investigations on CRIU to see if anything can be leveraged.

Just FYI, CRIU supports pre-copy live migration for containers.

**Comment by vmsearch** [comment]
@yisun-git  the private repo can not open

**Comment by yisun-git** [comment]
> @yisun-git the private repo can not open

Yes, it is private, only for internal work so far. :)

**Comment by vmsearch** [comment]
> > @yisun-git the private repo can not open
> 
> Yes, it is private, only for internal work so far. :)

Thanks your relply

**Comment by yisun-git** [comment]
Investigated possible migration states and list them here for discussion.

Basically, the states can be classified to two types.
1. External states: need interact with external components, e.g. KVM, to get/set states.
2. Internal states: maintained by cloud-hypervisor.

**External states**
![image](https://user-images.githubusercontent.com/22880141/65747225-7724e880-e133-11e9-84fd-c3c74f257840.png)

**Internal states**
![image](https://user-images.githubusercontent.com/22880141/65747756-ccadc500-e134-11e9-9703-e534570f881c.png)

**Comment by jiangliu** [comment]
The MigrationTransport trait should only provide interfaces to transport vmm state, not methods to create the channel (bind, connect etc).

**Comment by yisun-git** [comment]
> The MigrationTransport trait should only provide interfaces to transport vmm state, not methods to create the channel (bind, connect etc).

Thanks! It is for convenience to test in prototype codes. In formal codes, the non-common interfaces will not exist in the trait.

**Comment by wilsonwang371** [comment]
Hi Sun,

I am wondering how is the status of live migration now?

Wilson

**Comment by sboeuf** [comment]
@wilsonwang371 the live migration is not ready at the moment. We started merging some code related to the snapshot/restore functionalities, which the live migration will rely on, but there's still some work before we can start looking into live migration.

**Comment by rbradford** [comment]
This issue is obsolete.
